src/wallet/rpcdump.cpp

```diff
@@ -637,3 +638,425 @@ UniValue dumpwallet(const UniValue& params, bool fHelp)
+                        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Already have this key");
+                    }
+
+                    pwalletMain->mapKeyMetadata[vchAddress].nCreateTime = timestamp;
+
+                    if (!pwalletMain->AddKeyPubKey(key, pubkey)) {
+                        throw JSONRPCError(RPC_WALLET_ERROR, "Error adding key to wallet");
+                    }
+
+                    if (timestamp < pwalletMain->nTimeFirstKey) {
+                        pwalletMain->nTimeFirstKey = timestamp;
+                    }
+                }
+            }
+
+            success = true;
+        } else {
+            // Import public keys.
+            if (pubKeys.size() && keys.size() == 0) {
+                const string& strPubKey = pubKeys[0].get_str();
+
+                if (!IsHex(strPubKey)) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Pubkey must be a hex string");
+                }
+
+                std::vector<unsigned char> data(ParseHex(strPubKey));
+                CPubKey pubKey(data.begin(), data.end());
+
+                if (!pubKey.IsFullyValid()) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Pubkey is not a valid public key");
+                }
+
+                CBitcoinAddress pubKeyAddress = CBitcoinAddress(pubKey.GetID());
+
+                // Consistency check.
+                if (!isScript && pubKeyAddress.Get() != address.Get()) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Consistency check failed");
+                }
+
+                // Consistency check.
+                if (isScript) {
+                    CBitcoinAddress scriptAddress;
+                    CTxDestination destination;
+
+                    if (ExtractDestination(script, destination)) {
+                        scriptAddress = CBitcoinAddress(destination);
+                        if (scriptAddress.Get() != pubKeyAddress.Get()) {
+                            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Consistency check failed");
+                        }
+                    }
+                }
+
+                CScript pubKeyScript = GetScriptForDestination(pubKeyAddress.Get());
+
+                if (::IsMine(*pwalletMain, pubKeyScript) == ISMINE_SPENDABLE) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
+                }
+
+                pwalletMain->MarkDirty();
+
+                if (!pwalletMain->HaveWatchOnly(pubKeyScript) && !pwalletMain->AddWatchOnly(pubKeyScript)) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
+                }
+
+                // add to address book or update label
+                if (pubKeyAddress.IsValid()) {
+                    pwalletMain->SetAddressBook(pubKeyAddress.Get(), label, "receive");
+                }
+
+                // TODO Is this necessary?
```

Comment:
I was wondering the same when I checked the RPC call `importpubkey`, which calls the same _dance_ in [here](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcdump.cpp#L395).

In doubt, and since I did want to change the behavior of importing pub keys, I add the **TODO** mark for revision of this part of the code.